WatchFaceAlphaKotlin/app/src/main/java/com/example/android/wearable/alpha/data/WatchFaceUserSettingChangesDataSource.kt

```diff
@@ -0,0 +1,107 @@
+import com.example.android.wearable.alpha.utils.DRAW_HOUR_PIPS_STYLE_SETTING
+import com.example.android.wearable.alpha.utils.WATCH_HAND_LENGTH_STYLE_SETTING
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+
+/**
+ * Data Source for all changes by the user in settings related to the watch face styles. Coverts
+ * the watch face api callback to a flow.
+ */
+class WatchFaceUserSettingChangesDataSource(
+    private val currentUserStyleRepository: CurrentUserStyleRepository
+) {
+    @ExperimentalCoroutinesApi
+    fun getUserWatchFaceChanges(): Flow<WatchFaceUserChanges> = callbackFlow {
+        val userStyleChangeListener = object : CurrentUserStyleRepository.UserStyleChangeListener {
+            override fun onUserStyleChanged(userStyle: UserStyle) {
+                Log.d(TAG, "onUserStyleChanged(): userStyle: $userStyle")
+
+                var newWatchFaceUserChangesChanges: WatchFaceUserChanges = WatchFaceUserChanges()
+
+                for (options in userStyle.selectedOptions) {
+                    when (options.key.id.toString()) {
+                        COLOR_STYLE_SETTING -> {
+                            val listOption = options.value as
+                                    UserStyleSetting.ListUserStyleSetting.ListOption
+
+                            newWatchFaceUserChangesChanges = newWatchFaceUserChangesChanges.copy(
+                                colorId = listOption.id.toString()
+                            )
+                        }
+                        DRAW_HOUR_PIPS_STYLE_SETTING -> {
+                            val booleanValue = options.value as
+                                    UserStyleSetting.BooleanUserStyleSetting.BooleanOption
+
+                            newWatchFaceUserChangesChanges = newWatchFaceUserChangesChanges.copy(
+                                pipsEnabled = booleanValue.value
+                            )
+                        }
+                        WATCH_HAND_LENGTH_STYLE_SETTING -> {
+                            val doubleValue = options.value as
+                                    UserStyleSetting.DoubleRangeUserStyleSetting.DoubleRangeOption
+
+                            newWatchFaceUserChangesChanges = newWatchFaceUserChangesChanges.copy(
+                                minuteHandLength = doubleValue.value.toFloat()
+                            )
+                        }
+                        // TODO (codingjeremy): Add complication change support if settings activity
+                        // PR doesn't cover it. Otherwise, remove comment.
+                    }
+                }
+
+                try {
+                    trySend(newWatchFaceUserChangesChanges).isSuccess
+                } catch (error: Throwable) {
+                    Log.d(TAG, "error sending Flow: $error")
+                }
```

Comment:
Based on the usage and `UserStyleChangeListener` docs, I believe we only care about the most recent `UserStyle` since it is a complete snapshot (as opposed to partial updates).

If that's correct, we can add a `.buffer(Channel.CONFLATED)` right after the `callbackFlow`, which overrides the default, limited buffer size. Then `trySend` will only fail if the channel is closed, which indicates that we're being cancelled, so we can drop the `catch` block here.

Reply:
You are correct, I only care about the most recent version. 

For some reason, I'm getting an error when I add .buffer() after my callbackFlow, so I'll fix this in a followup commit tomorrow when I figure out what the issue is.